### DISCUSSION TITLE
How to override default parameters in tasks

### DISCUSSION THREAD

**Discussion by Ligtorn** [gaelcolas/Sampler#498] [Q&A] [answered]
I am trying to use the task publish_nupkg_to_gallery in release.module.build.ps1, but it needs a GalleryApiToken to run, as it has a condition on that parameter. I can see that script have that parameter, but I don't understand how I can provide the task with that parameters. 

I have tried setting it in build.yaml like this

```
  publish:
    - publish_nupkg_to_gallery  -GalleryApiToken 'xxx'
```
and     
```
  publish:
    - publish_nupkg_to_gallery  {-GalleryApiToken 'xxx' }

```      
I have also tried running it directly, without success

`.\build.ps1 -tasks publish -GalleryApiToken 'xxx' 
`

Unfortunately I fail to understand, how you can override the default parameters

I have tried searching the docs for a some clues in both this repo and in the invoke-build repo, but with no luck.

Can anybody guide me in the right direction

**Comment by johlju** [comment, marked as answer]
You pass in task script parameters either by passing them to `Invoke-Build`, set them as environment variables or set them as variables in the session prior to running `Invoke-Build`.

**Reply by Ligtorn** [reply]
It worked with environment variables, so that's great

If I understand you correctly, it should mean that I cannot set it via invoke-build, as sampler wraps that command via builds.ps1. Having to tweak that script would be a shame, as this is just the first of many modules I am creating

Or are there other ways of setting those parameters via builds.ps1 framework

**Reply by johlju** [reply]
Since this specific parameter is a sensitive value (a token) it wouldn't be a good idea to have in the build configuration. Also since it is an evaluation on the task to only run `publish_nupkg_to_gallery` if the value is set it would mean that the task would never run to be able to read it from the build configuration anyway.

To invoke the task using the build script the value have to be set prior to running the task. This can be done by another task that runs prior to the task `publish_nupkg_to_gallery` that reads the secret (token) from your secrets store and set an environment variable, or running something else prior to invoking the build script that sets the environment variable or variable is the current session.

There is currently not possible to pass those parameters to the build script. In my opinion that parameter is also to specific for a specific service to actually be a parameter on the build script. 🙂

**Reply by Ligtorn** [reply]
Thanks for the explanation, it makes a lot of sense to use the environment variables. This could be passed from a secret variable in the build pipeline, so it is protected properly
